File: Client/iTeamTalk/iTeamTalk/AppInfo.swift
/*
 * Copyright (c) 2005-2018, BearWare.dk
 * 
 * Contact Information:
 *
 * Bjoern D. Rasmussen
 * Kirketoften 5
 * DK-8260 Viby J
 * Denmark
 * Email: contact@bearware.dk
 * Phone: +45 20 20 54 59
 * Web: http://www.bearware.dk
 *
 * This source code is part of the TeamTalk SDK owned by
 * BearWare.dk. Use of this file, or its compiled unit, requires a
 * TeamTalk SDK License Key issued by BearWare.dk.
 *
 * The TeamTalk SDK License Agreement along with its Terms and
 * Conditions are outlined in the file License.txt included with the
 * TeamTalk SDK distribution.
 *
 */

import UIKit

class AppInfo {

    static let TTLINK_PREFIX = "tt://"

    static let OSTYPE = "iOS"

    static let DEFAULT_TCPPORT = 10333
    static let DEFAULT_UDPPORT = 10333
    
    static let WEBLOGIN_BEARWARE_USERNAME = "bearware"
    static let WEBLOGIN_BEARWARE_USERNAMEPOSTFIX = "@bearware.dk"
    static let WEBLOGIN_BEARWARE_PASSWDPREFIX = "token="
    
    static func isBearWareWebLogin(_ username: String) -> Bool {
        	return username == AppInfo.WEBLOGIN_BEARWARE_USERNAME ||
                username.hasSuffix(AppInfo.WEBLOGIN_BEARWARE_USERNAMEPOSTFIX)
    }
    
    enum BundleInfo {
        case name, version_NO, build_NO
    }
    
    static func getBundleInfo(_ b: BundleInfo) -> String {
        let bundle = Bundle.main
        let dict = bundle.infoDictionary

        switch b {
        case .name :
            if let info = dict?["CFBundleName"] {
                return info as! String
            }
            return "Unknown"
        case .version_NO :
            if let info = dict?["CFBundleShortVersionString"] {
                return info as! String
            }
            return "0.1"
        case .build_NO :
            if let info = dict?["CFBundleVersion"] {
                return info as! String
            }
            return "0"
        }
    }
    
    static func getAppName() -> String {
        return getBundleInfo(.name)
    }
    static func getAppVersion() -> String {
        return getBundleInfo(.version_NO)
    }
    static func getAppVersionLong() -> String {
        return getAppVersion() + " Build " + getBundleInfo(.build_NO)
    }

    static func getDefaultUrlArgs() -> String {
        return "client=" + getAppName() + "&version=" + getAppVersion() + "&dllversion=" + TEAMTALK_VERSION + "&os=" + OSTYPE
    }
    
    static func getServersURL(officialservers: Bool, unofficialservers: Bool) -> String {
        let officialstr = officialservers ? "1" : "0"
        let unofficialstr = unofficialservers ? "1" : "0"
        return "http://www.bearware.dk/teamtalk/tt5servers.php?official=\(officialstr)&unofficial=\(unofficialstr)&" + getDefaultUrlArgs()
    }

    static func getUpdateURL() -> String {
        return "http://www.bearware.dk/teamtalk/tt5update.php?" + getDefaultUrlArgs()
    }

    static var BEARWARE_REGISTRATION_WEBSITE = "http://www.bearware.dk"

    static func getBearWareTokenURL(username : String, passwd : String) -> String {
        
        let escUsername = username.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? username
        let escPasswd = passwd.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? passwd
        
        return "https://www.bearware.dk/teamtalk/weblogin.php?" + getDefaultUrlArgs() +
            "&service=bearware" + "&action=auth&username=" + escUsername + "&password=" + escPasswd
    }
    
    static func getBearWareServerTokenURL(username : String, token : String, accesstoken : String) -> String {
        
        let escUsername = username.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? username
        let escToken = token.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? token
        let escAccessToken = accesstoken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? accesstoken
        
        return "https://www.bearware.dk/teamtalk/weblogin.php?" + getDefaultUrlArgs() +
            "&service=bearware" + "&action=clientauth&username=" + escUsername +
            "&token=" + escToken + "&accesstoken=" + escAccessToken
    }
}

File: Client/TeamTalkAndroid/src/main/java/dk/bearware/data/AppInfo.java
/*
 * Copyright (c) 2005-2018, BearWare.dk
 * 
 * Contact Information:
 *
 * Bjoern D. Rasmussen
 * Kirketoften 5
 * DK-8260 Viby J
 * Denmark
 * Email: contact@bearware.dk
 * Phone: +45 20 20 54 59
 * Web: http://www.bearware.dk
 *
 * This source code is part of the TeamTalk SDK owned by
 * BearWare.dk. Use of this file, or its compiled unit, requires a
 * TeamTalk SDK License Key issued by BearWare.dk.
 *
 * The TeamTalk SDK License Agreement along with its Terms and
 * Conditions are outlined in the file License.txt included with the
 * TeamTalk SDK distribution.
 *
 */

package dk.bearware.data;

import android.content.Context;
import android.content.pm.PackageManager.NameNotFoundException;
import android.util.Log;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import dk.bearware.TeamTalkBase;

public class AppInfo {
    public static final String TAG = "bearware";

    public static final String APPNAME_SHORT = "TeamTalk5Droid";
    public static final String APPVERSION_POSTFIX = "";
    public static final String OSTYPE = "Android";

    public static final String WEBLOGIN_FACEBOOK_USERNAME = "facebook";
    public static final String WEBLOGIN_FACEBOOK_PASSWDPREFIX = "token=";

    public static final String WEBLOGIN_BEARWARE_USERNAME = "bearware";
    public static final String WEBLOGIN_BEARWARE_USERNAMEPOSTFIX = "@bearware.dk";

    // getUpdateUrl() response can override this
    public static String BEARWARE_REGISTRATION_WEBSITE = "http://www.bearware.dk";

    public static String getVersion(Context context) {
        String version = "";
        try {
            version = context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName;
        }
        catch(NameNotFoundException e) {
            Log.e(TAG, "Unable to get version information");
        }
        return version;
    }

    public static String getDefautlUrlArgs(Context context) {
        final String TEAMTALK_VERSION = TeamTalkBase.getVersion();
        String appversion = getVersion(context);
        return "client=" + APPNAME_SHORT + "&version="
                + appversion + "&dllversion=" + TEAMTALK_VERSION + "&os=" + OSTYPE;
    }

    public static String getServerListURL(Context context) {
        String urlToRead = "http://www.bearware.dk/teamtalk/tt5servers.php?" + getDefautlUrlArgs(context);
        return urlToRead;
    }

    public static String getUpdateURL(Context context) {
        String urlToRead = "http://www.bearware.dk/teamtalk/tt5update.php?" + getDefautlUrlArgs(context);
        return urlToRead;
    }

    public static String getBearWareTokenUrl(Context context, String username, String password) {

        try {
            username = URLEncoder.encode(username, "utf-8");
            password = URLEncoder.encode(password, "utf-8");
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "Unable to encode username/password: " + e.toString());
        }
        String urlToRead = "https://www.bearware.dk/teamtalk/weblogin.php?" + getDefautlUrlArgs(context) +
                "&service=bearware&action=auth&username=" + username + "&password=" + password;
        return urlToRead;
    }

    public static String getBearWareAccessTokenUrl(Context context, String username, String token, String accesstoken) {

        try {
            username = URLEncoder.encode(username, "utf-8");
            token = URLEncoder.encode(token, "utf-8");
            accesstoken = URLEncoder.encode(accesstoken, "utf-8");
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "Unable to encode username/token: " + e.toString());
        }
        String urlToRead = "https://www.bearware.dk/teamtalk/weblogin.php?" + getDefautlUrlArgs(context) +
                "&service=bearware&action=clientauth&username=" + username + "&token=" + token +
                "&accesstoken=" + accesstoken;
        return urlToRead;
    }

}
